src/sinks/clickhouse.rs

```diff
@@ -0,0 +1,295 @@
+    let in_flight_limit = config.request_in_flight_limit.unwrap_or(5);
+    let rate_limit_duration = config.request_rate_limit_duration_secs.unwrap_or(1);
+    let rate_limit_num = config.request_rate_limit_num.unwrap_or(5);
+    let retry_attempts = config.request_retry_attempts.unwrap_or(usize::max_value());
+    let retry_backoff_secs = config.request_retry_backoff_secs.unwrap_or(1);
+
+    let policy = FixedRetryPolicy::new(
+        retry_attempts,
+        Duration::from_secs(retry_backoff_secs),
+        HttpRetryLogic,
+    );
+
+    let uri = encode_uri(&host, &database, &table)?;
+
+    let http_service = HttpService::new(move |body: Vec<u8>| {
+        let mut builder = hyper::Request::builder();
+        builder.method(Method::POST);
+        builder.uri(uri.clone());
+
+        builder.header("Content-Type", "application/x-ndjson");
+
+        if gzip {
+            builder.header("Content-Encoding", "gzip");
+        }
+
+        builder.body(body).unwrap()
+    });
+
+    let service = ServiceBuilder::new()
+        .concurrency_limit(in_flight_limit)
+        .rate_limit(rate_limit_num, Duration::from_secs(rate_limit_duration))
+        .retry(policy)
+        .timeout(Duration::from_secs(timeout))
+        .service(http_service);
+
+    let sink = BatchServiceSink::new(service, acker)
+        .batched_with_min(
+            Buffer::new(gzip),
+            batch_size,
+            Duration::from_secs(batch_timeout),
+        )
+        .with(move |event: Event| {
+            let mut body = serde_json::to_vec(&event.as_log().all_fields()).unwrap();
+            body.push(b'\n');
+            Ok(body)
+        });
+
+    Ok(Box::new(sink))
+}
+
+fn healthcheck(host: String) -> super::Healthcheck {
+    // TODO: check if table exists?
+    let uri = format!("{}/?query=SELECT%201", host);
+    let request = Request::get(uri).body(Body::empty()).unwrap();
+
+    let https = HttpsConnector::new(4).expect("TLS initialization failed");
+    let client = Client::builder().build(https);
+    let healthcheck = client
+        .request(request)
+        .map_err(|err| err.to_string())
+        .and_then(|response| {
+            if response.status() == hyper::StatusCode::OK {
+                Ok(())
+            } else {
+                Err(format!("Unexpected status: {}", response.status()))
+            }
+        });
+
+    Box::new(healthcheck)
+}
+
+fn encode_uri(host: &str, database: &str, table: &str) -> Result<Uri, String> {
+    let query = url::form_urlencoded::Serializer::new(String::new())
+        .append_pair(
+            "query",
+            format!(
+                "INSERT INTO \"{}\".\"{}\" FORMAT JSONEachRow",
+                database,
+                table.replace("\"", "\\\"")
+            )
+            .as_str(),
+        )
+        .finish();
+
+    let url = if host.ends_with("/") {
+        format!("{}?{}", host, query)
+    } else {
+        format!("{}/?{}", host, query)
+    };
+
+    url.parse::<Uri>()
```

Comment:
Actually, I think host kind of works here since its combining the scheme and authority. While a Url or Uri (as used in the http sink) include a path and query. If we wanted to be technically correct we would name it scheme_and_authority, though that seems complicated.

The only reason I think we could change it to base_url is if its possible to talk to clickhouse with a `scheme://ip:port/path` type of url? Im not sure if this is even possible?

Reply:
Actually, it looks like ClickHouse doesn't use the path part of the URL at all, so both http://localhost:8123/?query=SELECT+1 and http://localhost:8123/any_path_here/?query=SELECT+1 return the same result. I can imagine someone using this feature to set up a load balancer/proxy that would route only requests certain path to ClickHouse and handle differently requests with other paths.